Plugin system and a generic way of representing errors

TobiasDeBruijn opened the pull request.

TobiasDeBruijn committed:
Implemented Plugin system and a generic way of representing errors

TobiasDeBruijn commented:
Just remembered I forgot one thing, don't merge just yet.

McTwist reviewed (approved):
While my expertise in Rust is low, I still understood from my knowledge how the system works and how it should be used. It could need some minor fixs and improvements in the future, but for now this will do.

TobiasDeBruijn committed:
Further work on plugin system
- Added tests for some functions
- Added declare_plugin! macro
- Changed expected type of plugin constructor to *mut $plugin_type, rather than *mut dyn Plugin, this is more FFI save and more idiomatic

TobiasDeBruijn commented:
Should be good to merge now. I've implemented tests on (I think) all places where they are doable to add. Some places do not have tests, notably the loading of libraries, as that is hard to unit-test, because it requires a precompiled unit. Integration tests for this maybe?
